### ISSUE TITLE
Permissions

### ISSUE THREAD

**Issue by universalhandle** [twomice/org.civicrm.extensionsui#9]
It is unclear what permissions are required to enable/disable/etc. extensions. I'd assumed it was _administer CiviCRM_, but this appears not to be the case.

Finding out what specific perms are relevant to extension interaction is a bit beyond the scope of this module, but ensuring that permissions-related failures are handled gracefully is not.

**Comment by twomice** [comment]
@GinkgoFJG Good point. Do you happen to have a tip for limiting access to Angular pages based on permissions?

**Comment by universalhandle** [comment]
I think the primary issue here is to determine what permission is required to mess with extensions (via API). I didn't dig deep into this, but I recall it seemed much more permissive than it ought to be (which obviously is a core issue, and not ours to deal with in this extension).

As access control is handled at the API level, our job is limited to providing a good user experience around that (i.e., don't show buttons that will result in an API-level permissions error), or perhaps a check when the page first renders (e.g., "Permission X is required to enable/disable extensions."). Of course we need to know what X is for this to be useful.

There isn't an established pattern that I'm aware for doing this. For CiviVolunteer I've used `CRM.checkPerm('do the thing')` on the client side. That method only works if CiviCRM has been told (server-side) which permissions it might be asked about on the client side. Prime the pump using `CRM_Core_Resources::singleton()->addPermissions(array('do the thing'))`.

**Comment by twomice** [comment]
I tried a couple of things to see about this:

Basic test: Call 
`CRM.api3('Extension', 'disable', {"keys": 'org.civicrm.module.cividiscount'})`
from a CiviCRM page to observe whether I can change the status of an extension.

1. As a user having neither "access ajax api" nor "administer CiviCRM":
Response is a 403.

2. As a user having only "access ajax api" but not "administer CiviCRM":
Response is:
```
{"error_code":"unauthorized","entity":"Extension","action":"disable","is_error":1,"error_message":"API
 permission check failed for Extension\/disable call; insufficient permission: require administer CiviCRM"
}
```

3. As a user having only "access ajax api" and "administer CiviCRM":
Response is:
`{"is_error":0,"version":3,"count":1,"values":1}`

**Conclusion:**
Based on the error message in item 2 (above) it does look like "administer CiviCRM" is required to modify the status of extensions. I think we're okay here.

**Comment by universalhandle** [comment]
For due (or excessive?) diligence, I went ahead and created myself a user with "access CiviCRM" but not "administer CiviCRM," just to see how gracefully the permissions failures are handled.

The first thing to note is that the menu item does not appear, since the "Administer" menu and subitems are not available to me. Great.

Next I punched in the URL directly and got this (see image below). Nothing that shouldn't be is given away. I can see that this is a screen where I should be able to manage extensions, but there's nothing for me to actually do. There's no "access denied" message or friendly "are you lost?" message, but I can live with that since the only way someone is getting to this screen without the appropriate permissions is if they're up to no good.
![unpermissioned](https://user-images.githubusercontent.com/871819/34011315-a2b96564-e0dd-11e7-93a2-310daa8abb1b.png)

Clicking the "refresh" button gives an appropriate error message.
![click-refresh](https://user-images.githubusercontent.com/871819/34011321-a854a466-e0dd-11e7-89e0-d79b2a653af4.png)

I'm satisfied!
